src/generators/utils/parse.ts

```diff
@@ -47,3 +55,132 @@ export function parseSitesFixesConfig<T extends SiteProps>(text: string, options
+    return offsets.map(([offset, length]) => {
+        const stringOffset = offset.toString(36);
+        const stringLength = length.toString(36);
+        return '0'.repeat(4 - stringOffset.length) + stringOffset + '0'.repeat(3 - stringLength.length) + stringLength;
+    }).join('');
+}
+
+function decodeOffset(offsets: string, index: number): [number, number] {
+    const base = (4 + 3) * index;
+    const offset = parseInt(offsets.substring(base + 0, base + 4), 36);
+    const length = parseInt(offsets.substring(base + 4, base + 4 + 3), 36);
+    return [
+        offset,
+        offset + length,
+    ];
+}
+
+export function indexSitesFixesConfig<T extends SiteProps>(text: string): SitePropsIndex<T> {
+    const domains: {[domain: string]: number | number[]} = {};
+    const domainPatterns: {[domainPattern: string]: number | number[]} = {};
+    // Array of tuples, where first number is an offset of record start and second number is record length.
+    const offsets: Array<[number, number]> = [];
+
+    function processBlock(recordStart: number, recordEnd: number, index: number) {
+        // TODO: more formal definition of URLs and delimiters
+        const block = text.substring(recordStart, recordEnd);
+        const lines = block.split('\n');
+        const commandIndices: number[] = [];
+        lines.forEach((ln, i) => {
+            if (ln.match(/^[A-Z]+(\s[A-Z]+){0,2}$/)) {
+                commandIndices.push(i);
+            }
+        });
+
+        if (commandIndices.length === 0) {
+            return;
+        }
+
+        const urls = parseArray(lines.slice(0, commandIndices[0]).join('\n'));
+        for (const url of urls) {
+            const domain = getDomain(url);
+            if (isFullyQualifiedDomain(domain)) {
+                if (!domains[domain]) {
+                    domains[domain] = index;
+                } else if (typeof domains[domain] === 'number' && domains[domain] !== index) {
+                    domains[domain] = [(domains[domain] as number), index];
+                } else if (typeof domains[domain] === 'object' && !((domains[domain] as number[]).includes(index))) {
+                    (domains[domain] as number[]).push(index);
+                }
+                continue;
+            }
+
+            if (!domains[domain]) {
+                domainPatterns[domain] = index;
+            } else if (typeof domains[domain] === 'number' && domains[domain] !== index) {
+                domainPatterns[domain] = [(domainPatterns[domain] as number), index];
+            } else if (typeof domains[domain] === 'object' && !((domains[domain] as number[]).includes(index))) {
+                (domainPatterns[domain] as number[]).push(index);
+            }
+        }
+        offsets.push([recordStart, recordEnd - recordStart]);
+    }
+
+    let recordStart = 0;
+    // Delimiter between two blocks
+    const delimiterRegex = /\r?\s*={2,}\s*\r?/gm;
```

Comment:
Hmmm, maybe we should ban the `\r`?

Reply:
I'd gladly remove `\r` from the code, since it is an archaic character not relevant in modern systems. I added this because current code permits source file to have `\r` (it just removes them):
https://github.com/darkreader/darkreader/blob/9c664ff170841c46322424795098e61629919f8a/src/generators/utils/parse.ts#L16

Perhaps, we could just forbid `\r` in configs (and add a unit test and throw a warning if user edits config and introduces this character).

Reply:
I can live with that.

Reply:
Please see #6690.